### PR TITLE
[Rewrite] Support for 2.6, 3.2 and pypy

### PR DESCRIPTION
**Pull request by ionelmc**
Now there are only failures on pypy3 (which looks like a bug in setuptools or pip).

I've also extended the integration tests to run with all target pythons. That will combine with all the tox envs (the host).

### CODE REVIEW

**Comment by ionelmc** [comment]
Note that this includes #702

**Comment by dstufft** [comment]
I'm taking a look at this, I'm probably going to end up pulling it in in pieces so I can make sure I fully understand all the parts.

File: virtualenv/_utils.py [hunk 1]
```diff
@@ -1,25 +1,34 @@
...
-    # We use copyfile (not move, copy, or copy2) to be extra sure that we are
-    # not moving directories over (copyfile fails for directories) as well as
-    # to ensure that we are not copying over any metadata because we want more
-    # control over what metadata we actually copy over.
-    shutil.copyfile(srcfile, destfile)
+def copyfile(srcfile, destfile, skip=re.compile(r".*\.pyc\Z|__pycache__\Z", re.IGNORECASE)):
+    ensure_directory(os.path.dirname(destfile))
+
+    if os.path.isdir(srcfile):
+        # TODO: just use shutil.copytree to avoid bikeshedding
```
Comment (dstufft) [inline virtualenv/_utils.py]: What do you mean by "avoid bikeshedding"?

Comment (ionelmc) [inline virtualenv/_utils.py]: So this seems to me very similar to what we have in Python's `shutil`. I wrote that comment to draw attention and prompt discussion about whether there's a better way to handle copying.

It could use a better name at least but I couldn't decide.

Comment (dstufft) [inline virtualenv/_utils.py]: The reason it's using this code instead of what's in ``shutil`` is because the code in ``shutil`` copies a bunch of extra metadata which we don't want to copy. In particular things like owner/group, SELinux contexts and the like. We want the copied files to act like "new" files that just happen to have the same content.

File: virtualenv/builders/legacy.py [hunk 2]
```diff
@@ -175,54 +208,94 @@ def create_virtual_environment(self, destination):
...
+
+        dst = os.path.join(os.path.dirname(osmoduledestination), "site.py")
+        logger.debug("Writing %s", dst)
+        with io.open(dst, "wb") as dst_fp:
             # Get the data from our source file, and replace our special
             # variables with the computed data.
             data = SITE
-            data = data.replace("__PREFIX__", repr(destination))
-            data = data.replace("__EXEC_PREFIX__", repr(destination))
+            data = data.replace(b"__PREFIX__", self._path_repr(destination))
```
Comment (ionelmc) [inline virtualenv/builders/legacy.py]: If anyone is wondering about the weird `_path_repr` it's just my poor attempt to deal with various string/bytes literals over all the pythons. I still need to add integration tests with weird unicode chars in them.

File: virtualenv/flavors/base.py [hunk 3]
```diff
@@ -6,24 +6,95 @@
 
 class BaseFlavor(object):
 
-    core_modules = set([
-        "posixpath.py", "stat.py", "genericpath.py", "warnings.py",
-        "linecache.py", "types.py", "UserDict.py", "_abcoll.py", "abc.py",
-        "_weakrefset.py", "copy_reg.py",
-    ])
+    def bootstrap_modules(self, python_info):
```
Comment (ionelmc) [inline virtualenv/flavors/base.py]: Only the available modules or directories (note: *not necessarily packages*) will be copied.

File: virtualenv/builders/venv.py [hunk 4]
```diff
@@ -53,4 +53,4 @@ def create_virtual_environment(self, destination):
         # We use the programatic API instead of the command line because we
         # want to not include pip in this, and Python 3.3's venv module doesn't
         # support the --without-pip flag.
-        subprocess.check_call([self.python, "-c", script])
+        subprocess.check_call([self._python_bin, "-c", script])
```
Comment (ionelmc) [inline virtualenv/builders/venv.py]: So yes, sadly, Python 3.3's `venv` blows up if used from a virtualenv.

File: virtualenv/_utils.py [hunk 5]
```diff
@@ -1,25 +1,34 @@
...
-import errno
 import os
+import re
 import shutil
 import stat
 
 
 def ensure_directory(directory, *args, **kwargs):
-    try:
+    # Fail if the destination exists and it's not a directory
```
Comment (ionelmc) [inline virtualenv/_utils.py]: So no, this is not merely a style change. This is to bail early when due to some misconfiguration or bug the directory copying from the builder would attempt to copy a directory when there already a file with the same name in the destination.

**Comment by sigmavirus24** [comment]
Holy pull request batman. 26 changed files. Over 900 changed lines. This is a monster. @ionelmc have you thought of splitting this up into smaller, easier to review PRs?

**Comment by ionelmc** [comment]
@sigmavirus24 think of it like the rewrite of the rewrite :-)

**Comment by pfmoore** [comment]
Wow. I think it'll have to be pulled apart before it's committed. @dstufft above said he'd want to pull it to pieces, and I agree, although I'd rather see independent PRs than bits cherry-picked from here. While I've not been following the rewrite much, I can't even see how I'd begin to review this.

If you burn @dstufft out, I'll be cross with you ;-)

File: tests/functional/test_create.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+            return binpath
+
+
+def resolve_path(path):
+    if path is not None:
+        path = os.path.realpath(os.path.expanduser(path))
+    return path
+
+
+PYTHON_BINS = [
```
Comment (Carreau) [inline tests/functional/test_create.py]: I would add a comment of what each of these value are. Like a header for an array. The second and third are obisou, not the first.

File: tests/functional/test_create.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    (locate_on_path("pypy"), None),
+]:
+    # I'm terrible here but I want certain checks disabled for these paths: the --system-site-packages checks, otherwise
+    # I have to reimplement bin resolving here, and it's a bad idea to duplicate logic in tests.
+    if (True, path, sitepackages) not in PYTHON_BINS:
+        if path:
+            PYTHON_BINS.append((False, path, sitepackages))
+
+PYTHON_BINS = [
+    (is_global, path, sitepackages)
```
Comment (Carreau) [inline tests/functional/test_create.py]: could we apply resolve-path inconditionaly here instead of line 41-47 ? (not sure how it behave on windows, just proposing)

File: tests/functional/test_tox.ini [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+basepython = {env:TOXPYTHON:python2.7}
+
+[testenv:py32]
+basepython = {env:TOXPYTHON:python3.2}
+
+[testenv:py33]
+basepython = {env:TOXPYTHON:python3.3}
+
+[testenv:py34]
+basepython = {env:TOXPYTHON:python3.4}
```
Comment (Carreau) [inline tests/functional/test_tox.ini]: You seem to be missing `[testenv:pypy]` and `[testenv:pypy3]` or are they unnecessary ?

File: tests/functional/test_create.py (hunk 7 above)
Comment (ionelmc) [inline tests/functional/test_create.py]: I could move the `resolve_path` down here but then this would misbehave:
```
if (True, path, sitepackages) not in PYTHON_BINS:
```

File: tests/functional/test_tox.ini [hunk 9]
```diff
@@ -0,0 +1,25 @@
+[tox]
+envlist = py26,py27,pypy,py32,py33,py34,pypy3
```
Comment (Carreau) [inline tests/functional/test_tox.ini]: if possible, I would for later reader keep the version in the same order. Sometime pypy is before Python2, sometime both pypy and pypy3 are at the end.

File: tests/functional/test_create.py (hunk 7 above)
Comment (Carreau) [inline tests/functional/test_create.py]: Ah, indeed.

File: tests/functional/test_tox.ini (hunk 8 above)
Comment (ionelmc) [inline tests/functional/test_tox.ini]: Yes indeed you are correct, they are not there cause I didn't add PyPy into the AppVeyor test grid yet. It's on my todo list

File: circle.yml [hunk 10]
```diff
@@ -0,0 +1,6 @@
+test:
+  pre:
+  - ls -al ~/.pyenv/shims ~/.pyenv/versions 
+  - pyenv global 2.6.8 2.7.9 pypy-2.4.0 3.2.5 3.3.3 3.4.1
```
Comment (Carreau) [inline circle.yml]: CircleCI has no `pypy-3`?

File: tests/functional/test_tox.ini (hunk 8 above)
Comment (Carreau) [inline tests/functional/test_tox.ini]: Ok, fair enough.

File: tests/functional/test_tox.ini (hunk 9 above)
Comment (ionelmc) [inline tests/functional/test_tox.ini]: Hmmm, they have the same other as in the main tox.ini (by python version). Should I reorder?

File: circle.yml (hunk 10 above)
Comment (ionelmc) [inline circle.yml]: This is all they have now:
```
/home/ubuntu/.pyenv/versions:
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.6.6
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:22 2.6.8
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:23 2.7
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.7.3
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.7.4
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.7.5
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.7.6
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:21 2.7.7
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:22 2.7.8
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:22 2.7.9
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:24 3.1.5
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:22 3.2
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:24 3.2.5
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:22 3.3.0
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:23 3.3.2
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:23 3.3.3
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:23 3.4.0
drwxr-xr-x 1 ubuntu ubuntu  42 Jan  1 15:23 3.4.1
drwxr-xr-x 1 ubuntu ubuntu 122 Jan  1 15:24 pypy-2.2.1
drwxr-xr-x 1 ubuntu ubuntu 122 Jan  1 15:24 pypy-2.3.1
drwxr-xr-x 1 ubuntu ubuntu 122 Jan 1 15:24 pypy-2.4.0
```

Comment (Carreau) [inline circle.yml]: Ok, no problem. I'm just trying to do a minimal review to lighten the burden of repo maintainer.

**Comment by gaborbernat** [comment]
I'll close this now as we no longer support those. Feel free to open new PRs for features present here but not on master.
